### ISSUE TITLE
Remove `void{}` from the language

### ISSUE THREAD

**Issue by mlugg** [ziglang/zig#15213]
## TLDR

- `void{}` is unused in all real code
- `void{}` is a pointless second way of achieving the same thing as `{}`
- `void{}` is a special case which prevents `T{}` syntax from being for aggregate initialization only
- `void{}` should therefore be removed from the language.

## Not Long Enough, Let Me Read

Currently, there are two ways to write a `void` literal: `{}`, and `void{}`.

`{}` is the canonical form: it's explicitly recommended over `void{}`, works in every case `void{}` does, and is used in all Zig code I've ever seen. The form `void{}` does not exist in the entire compiler and standard library, and I've never seen it used in the wild. There seems to be no practical reason for it to exist.

Despite its lack of use, I'd argue its existence still violates the Zen of `Only one obvious way to do things`. To someone not explicitly familiar with Zig's conventions, it's not at all clear which is preferable: maybe explicitness is preferred?

The form `void{}` also leads to an inconsistency regarding initialization syntax. In general, it is a rule that `.{}` is an *aggregate* initializer, not a general type initializer (you can't initialize a `u32`, a `*T`, etc with it). That's why, in stage2, it was made that `.{}` is no longer a valid initializer for `void` values. However, in every other case in the language, it is the case that `.{}` is precisely equivalent to `T{}` where `T` is the type being initialized; therefore, we can also say that `T{}` for any `T` is aggregate initialization syntax. Since `void` is not an aggregate, maintaining the `void{}` syntax is inconsistent, particularly with the removal of `.{}` initialization for `void` (which is, for the record, a decision I agree with).

The `void{}` form should be eliminated from the language. A `zig fmt` fixup would be nice, but honestly isn't even really necessary IMO since I've genuinely never seen anyone use this form.

**Comment by rohlem** [comment]
I've been using it, but I'll stop now.
If someone really wants to keep using it, they can carry on with `const Void = struct{};`.

**Comment by Jarred-Sumner** [comment]
> void{} is unused in all real code
>
> The void{} form should be eliminated from the language. A zig fmt fixup would be nice, but honestly isn't even really necessary IMO since I've genuinely never seen anyone use this form.

I can find at least 190 real uses of `void{}`:

<img width="135" alt="image" src="https://user-images.githubusercontent.com/709451/230772372-08b34ffe-ac37-470f-abb1-a7e3f3cdf198.png">

That being said, makes sense to remove it imo (and it's easy to find-and-replace)

**Comment by nektro** [comment]
i generally support this proposal but i will add that fwiw `void{}` is more grep-able than `{}`

**Comment by r00ster91** [comment]
It doesn't really matter how grepable it is if nobody uses it.

**Comment by mlugg** [comment]
More to the point, when would you ever need to grep for void constants?

**Comment by mlugg** [comment]
@Jarred-Sumner

> just to play devil's advocate:
> 
> which code snippet is easier to read?
> 
> ```zig
> std.sort.sort(
>     WatchItemIndex,
>     evict_list[0..evict_list_i],
>     void{},
>     std.sort.desc(WatchItemIndex),
> );
> ```
> 
> ```zig
> std.sort.sort(
>     WatchItemIndex,
>     evict_list[0..evict_list_i],
>     {},
>     std.sort.desc(WatchItemIndex),
> );
> ```

I find the second snippet easier to read (in the sense that I understand it faster); when I was less experienced with Zig, I suspect I would have had no strong preference.

**Comment by sigod** [comment]
I would argue in favor of removing `{}` and keeping `void{}`. Because of:
- Consistency. `void{}` is consistent with the syntax for other types, e.g. `Foo{}`. Why have a different syntax for `void`?
- Newcomer friendliness. I remember being confused when seeing `{}` for the first time. It makes one think that it was supposed to be `.{}`, but a dot was lost. It's basically a magic syntax, you have to know that the result is `void`.
- grep-ability. As already mentioned.

> which code snippet is easier to read?

The first one.

**Comment by mlugg** [comment]
> - Consistency. `void{}` is consistent with the syntax for other types, e.g. `Foo{}`. Why have a different syntax for `void`?

As addressed in the original issue post, this syntax is inconsistent with the rest of the language, since `T{}` is otherwise an *aggregate* initializer.

> - Newcomer friendliness. I remember being confused when seeing `{}` for the first time. It makes one think that it was supposed to be `.{}`, but a dot was lost. It's basically a magic syntax, you have to know that the result is `void`.

How is `{}` any more of a "magic syntax" than `.{}`? It's just *a* syntax, which you need to learn - it's pretty simple and is easy to remember once you know what it is. Having to learn some basic bits of syntax is kind of just how programming languages work.

> - grep-ability. As already mentioned.

As I previously asked: why does this actually matter? When do you need to grep your codebase for `void` values?

**Comment by rohlem** [comment]
> It's basically a magic syntax

@sigod The syntax arises naturally from being an empty unlabeled block `{}`.
In status-quo you can already write a block of instructions as any expression, though it has to be labeled if you intend to return a non-void value:
```zig
const std = @import("std");
const a: u32 = 8;
const b: u32 = b: { // constructed using a block instead of just a literal
    // arbitrary statements
    std.debug.assert(10 > 5);
    break :b 8;
};
const c: void = c: { // now we provide `void` instead of `u8` as value
    std.debug.assert(10 > 5);
    break :c void{};
};
const d: void = { // providing no value implicitly provides a void value
    std.debug.assert(10 > 5);
    // comparable to how `void` functions aren't required to have a `return` statement
};
const e: void = {}; // same as above, but empty (has no statements).
const f = {}; //same as e (`void` deduced.
```

And as I've pointed out before, if you prefer the syntax `void{}` and want to enforce it in your own code:
```zig
pub const Void = struct{}; // use this instead of `void` and now you _have to use `Void{}` instead of `{}` everywhere!
```

**Comment by KilianHanich** [comment]
I am not sure, but wouldn't this mean that this code wouldn't compile anymore if `void{}` was removed:

```Zig
fn f(comptime T: type) type {
    // do something interesting
    return struct {
        marker: T = T{},
        // other stuff
    };
}

const my_f = f(void);
```

Sure, `f(i32)` doesn't work (since it doesn't support array initialization syntax, at least that's the compiler error), but sometimes you use a void value in a generic structure as a placeholder in case you don't need that compiler for your specific case while in other places you do.

**Comment by kj4tmp** [comment]
I am a newcomer (< 3 months), so I will not provide an opinion on the proposal, but some have wondered what newcomers are thinking, so this is what I am thinking:

1. I view `{}` as brackets. I use them to say "this is where this function begins and ends" and I am most concerned with having the right number of opening and closing brackets. I did not know it was an instance of `void`.
2. I use `void` to tell the compiler that my function will not return anything, I don't know much beyond that yet.
3. Coming from python, this issue made me realize that functions annotated to return ` -> None` should actually be annotated to return `NoneType` and now I can never unsee this inconsistency :)
4. Also coming from python, I know `None` is a singleton and a keyword. Which begs the question of maybe zig should have a keyword for the single void instance. After reading a lot of tickets I would expect zig to not add a keyword because zig is very focused on being a simple language. But as a newcomer I see `{}` as effectively an unreadable keyword for an instance of void.
5. I read `void{}` as "an instance of void" because I use {} to make instances of things.

The concept of _aggregate initializers_ was difficult for me to understand at first because I thought "actually I can initialize anything with T{}!"
 because I did not fundamentally understand this array initialization:
```zig
const std = @import("std");

test {
    const my_array: [4]u8 = [_]u8{1} ** 4;
    try std.testing.expectEqual(.{ 1, 1, 1, 1 }, my_array);
}
```

My mental model was "I'm making 4 bytes here" when it should have been "I am concatenating 4 single-item arrays of bytes to make 1 array of 4 bytes". Once I understood that I was initializing single item arrays, I understood that this was an _aggregate initialization_.
(Perhaps the example in the langref should change to be `[1]u8{1} ** 4;` and not use the fancy auto-assume array length feature).


As I become more advanced and write more generic code, I could see the use case highlighted by @KilianHanich being annoying, as I would have to pass an empty struct instead of void to represent "nothing", and thus know the difference between aggregate initializers and whatever the other initializers are called.

**Comment by kj4tmp** [comment]
I would also mention that learning is path dependent.

In my case, I learned in the following order:
1. how to make a function that returns nothing (using `void`) (when writing `pub fn main`)
3. how to make a struct type (when I first needed to return multiple things from a function)
4. how to instantiate a struct (using `.{}` and `Foo{}`) (to create the return value of that function)
5. how to make an instance of void (from this issue, using void{}) which naturally follows from 4.

**Comment by pfgithub** [comment]
@KilianHanich Correct, that code would no longer work. You would have to update it to know the default value or provide the value on initialization

```diff
-fn f(comptime T: type) type {
+fn f(comptime T: type, comptime default: T) type {
    // do something interesting
    return struct {
-       marker: T,
+       marker: T = default,
        // other stuff
    };
}

-const my_f = f(void);
+const my_f = f(void, {});
```

**Comment by rohlem** [comment]
@KilianHanich Special-casing inline is also an option - since `type` values are comptime-only, the condition is `comptime`-known and only the taken branch will be analyzed:
```diff
-       marker: T,
+       marker: T = if (T == void) {} else T{},
```

**Comment by KilianHanich** [comment]
@rohlem Yes, but while reading generic code you normally don't want to think about stuff like that if it isn't semantically (as in what's going on with the code) needed, and not just work around what is essentially a syntax annoyance (which is also the reason why I am of the opinion that `.{}` should work for all types, but I guess that's a different topic too).

**Comment by andrewrk** [comment]
https://codeberg.org/ziglang/zig/commit/5d55999d20c4d856ad9e0bd35a939e4aa3be3d0d
